prog/dump/i2cset.8
.TH I2CSET 8 "June 2007"
.SH "NAME"
i2cset \- set I2C registers

.SH SYNOPSIS
.B i2cset
.RB [ -f ]
.RB [ -y ]
.I i2cbus
.I chip-address
.I data-address
.I value
.RI [ mode ]
.RI [ mask ]
.br
.B i2cset
.B -V

.SH DESCRIPTION
i2cset is a small helper program to set registers visible through the I2C
bus.

.SH OPTIONS
.TP
.B -V
Display the version and exit.
.TP
.B -f
Force access to the device even if it is already busy. By default, i2cset
will refuse to access a device which is already under the control of a
kernel driver. Using this flag is dangerous, it can seriously confuse the
kernel driver in question. It can also cause i2cset to silently write to
the wrong register. So use at your own risk and only if you know what
you're doing.
.TP
.B -y
Disable interactive mode. By default, i2cset will wait for a confirmation
from the user before messing with the I2C bus. When this flag is used, it
will perform the operation directly. This is mainly meant to be used in
scripts.
.PP
There are four required options to i2cset. \fIi2cbus\fR indicates the number
of the I2C bus to be scanned.  This number should correspond to one of
the busses listed by \fIi2cdetect -l\fR. \fIchip-address\fR specifies the
address of the chip on that bus, and is an integer between 0x00 and 0x7F.
\fIdata-address\fR specifies the address on that chip to write to, and is an
integer between 0x00 and 0xFF. \fIvalue\fR is the value to write to that
location on the chip.
.PP
The \fImode\fR parameter, if specified, is one of the letters \fBb\fP or
\fBw\fP, corresponding to a write size of a single byte or a 16-bit word,
respectively. A \fBp\fP can also be appended to the \fImode\fR parameter to
enable PEC. If the \fImode\fR parameter is omitted, i2cset defaults to byte
mode without PEC. The \fIvalue\fR provided must be within range for the
specified data type (0x00-0xFF for bytes, 0x0000-0xFFFF for words).
.PP
The \fImask\fR parameter, if specified, describes which bits of \fIvalue\fR
will be actually written to \fIdata-address\fR. Bits set to 1 in the mask
are taken from \fIvalue\fR, while bits set to 0 will be read from
\fIdata-address\fR and thus preserved by the operation.

.SH WARNING
i2cset can be extremely dangerous if used improperly. It can confuse your
I2C bus, cause data loss, or have more serious side effects. Writing to
a serial EEPROM on a memory DIMM (chip addresses between 0x50 and 0x57) may
DESTROY your memory, leaving your system unbootable!  Be extremely careful
using this program.

.SH SEE ALSO
i2cdump(8), isaset(8)

.SH AUTHOR
Frodo Looijaard, Mark D. Studebaker, and the lm_sensors group
http://www.lm-sensors.org/
This manual page was originally written by David Z Maze <dmaze@debian.org> for
the Debian GNU/Linux system. It was then reviewed by the lm_sensors team and
is now part of the lm_sensors source distribution.


CHANGES
lm_sensors CHANGES file
-----------------------

SVN HEAD
  File i2c-dev.h: Support I2C block reads with specified length
  File sensors.conf.eg: Add a dme1737 section
  File lm_sensors.sysconfig: Delete, now generated by sensors-detect
  Library: Clean up pwm symbol names translation
           Add dme1737 support
           Add applesmc support
           Add w83627ehf support (for Linux 2.4)
           Add support for the w83627ehf VID function
           Add support for the w83627ehf thermal sensor types
  Man page sensors.conf.5: Update the chip statement section
  Module i2c-nforce2: Add nForce MCP61, MCP65 support (2.6 backport)
  Module w83627ehf: New (2.6 backport)
  Programs i2cdump, i2cget, i2cset: Do not force the slave address by default
  Program i2cdump: Use the new I2C block read function
  Program isadump: Detect when address bit 7 is a busy flag
                   Fix Super-I/O exit sequence for Winbond/Fintek chips
  Program fancontrol: Use let for arithmetic evaluation
                      No longer need awk and grep
                      Limit calls to external programs (Linux 2.6 only)
                      Check for configuration file validity
                      Support optional min and max PWM values
                      Better integration with init scripts
  Program pwmconfig: Use smaller steps for low PWM values
                     Support optional min and max PWM values
  Program mkpatch: Include w83627ehf
  Program rrd: Support non-i2c devices (hwmon class)
               Add missing shell declarations
               Update URI
  Program sensors: Drop ddcmon and eeprom support
                   Add dme1737 support
                   Add applesmc support
                   Add max6680 support
                   Fix alignment of alarm for one-limit temperatures
                   Add w83627ehf support (for Linux 2.4)
                   Add support for the w83627ehf VID function
                   Add support for the w83627ehf thermal sensor types
  Program sensors-detect: Stop Super-I/O probe after first family success
                          Fix SMSC DME1737 detection
                          Add /usr/sbin to the PATH (#2199)
                          Add ATI SB700 detection
                          Add Maxim MAX6680/MAX6681 detection
                          Add IT8726F detection
                          Fix IPMI support for 2.6 kernels
                          Add detection for non-standard SMSC Super-I/Os
                          No longer depend on i2cdetect
                          Add SMSC SCH5317 detection
                          Drop detection of most non-sensors I2C chips


2.10.3 (20070319)
  Library: Do not skip i2c adapters with only a class device and no
            physical device. This change is required to survive the
            planned struct class_dev removal from future 2.6 kernels.
           Add support for the SMSC LPC47M292 fans (prefix smsc47m2)
           Recognize i2c-isa-based devices as such even after class_dev
	    is gone
           Compile with -D_REENTRANT
           Fix support of Linux 2.6's max6650 driver
           Fix max1619 support
  Man page i2cdetect.8: Scanning range can be restricted
  Module bmcsensors: Fix debugging messages
  Module i2c-ali1563: Improve the status messages (2.6 backport)
                      Fix device initialization (2.6 backport)
  Module i2c-amd8111: Cleanups (2.6 backport)
                      Fix initialization race (2.6 backport)
  Module i2c-nforce2: Drop unused reference to pci_dev (2.6 backport)
  Module i2c-piix4: Add ATI SB600 support (2.6 backport)
  Module i2c-viapro: Add CX700 support (2.6 backport)
  Module smsc47m1: Add SMSC LPC47M292 support
                   Get rid of a useless semaphore (2.6 backport)
  Programs i2cdetect, i2cdump, isadump: Flush output in real time
  Program i2cdump: Cleaner output in SMBus block read mode
  Program fancontrol.pl: Fix pwmN_enable access (#2190)
  Program sensors: Add smsc47m2 support (fans)
                   Fix -A option (hide adapter name) on error
  Program sensors-detect: Only probe relevant I2C addresses
                          Drop ARP-capable device detection
                          Add Maxim MAX6655/MAX6656 detection
                           (Olexiy Avramchenko)
                          Add Maxim MAX1668, MAX1805, MAX1989 detection
                          Add VIA CX700 detection
                          Drop legacy W83627HF ISA detection
                          Don't bufferize stdout during probes
                          Speed up the LM75 and LM77 detection
                          Clean up the EEPROM detection
                          Add Dallas DS75 detection
                          Add SMSC LPC47M292 detection
                          Add nVidia MCP61 and MCP65 detection
                          Fix ADM1028 detection
                          Add Nat. Semi. PC8347L Super-I/O detection
                          Add SMSC SCH5504 Super-I/O detection (no sensors)
                          Make the LM78 and W83781D detection safer


2.10.2 (20070115)
  RPM: Obsolete, removed
  File doc/fan-divisors: Update
  File sensors_vid.h: Trim VID values to correct number of bits (2.6 backport)
  Library: Add support for the pc87247 driver (fans only)
           Probe for busses before chips
           Drop support for algorithm names
           Add support for coretemp driver   
           Major update/optimization of config-file scanner
           Add config-file scanner regression tests
           Add f71872f support
           Add w83627dhg support (David Holl, #2157)
           Get the i2c adapter names from classdev.name
           Add adm1029 support (Corentin Labbe)
  Man page i2cdetect.8: Describe the output convention
  Man page sensors.1: Update (option -c) and clean up
  Module bmcsensors: Fix several warnings
                     Fix fragile structure initialization

prog/dump/i2cset.c
/*
    i2cset.c - A user-space program to write an I2C register.
    Copyright (C) 2001-2003  Frodo Looijaard <frodol@dds.nl>, and
                             Mark D. Studebaker <mdsxyz123@yahoo.com>
    Copyright (C) 2004-2005  Jean Delvare <khali@linux-fr.org>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include <errno.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include "i2cbusses.h"
#include "util.h"
#include "i2c-dev.h"
#include "version.h"

void help(void) __attribute__ ((noreturn));

void help(void)
{
	fprintf(stderr, "Syntax: i2cset [-f] [-y] I2CBUS CHIP-ADDRESS DATA-ADDRESS "
	        "VALUE [MODE] [MASK]\n"
	        "        i2cset -V\n"
	        "  MODE is 'b[yte]' or 'w[ord]' (default b)\n"
	        "  Append 'p' to MODE for PEC checking\n"
	        "  I2CBUS is an integer\n");
	print_i2c_busses(0);
	exit(1);
}

int main(int argc, char *argv[])
{
	char *end;
	int res, i2cbus, address, size, file;
	int value, daddress, vmask = 0;
	int e1;
	char filename[20];
	unsigned long funcs;
	int pec = 0;
	int flags = 0;
	int force = 0, yes = 0, version = 0;

	/* handle (optional) flags first */
	while (1+flags < argc && argv[1+flags][0] == '-') {
		switch (argv[1+flags][1]) {
		case 'V': version = 1; break;
		case 'f': force = 1; break;
		case 'y': yes = 1; break;
		default:
			fprintf(stderr, "Warning: Unsupported flag "
				"\"-%c\"!\n", argv[1+flags][1]);
			help();
			exit(1);
		}
		flags++;
	}

	if (version) {
		fprintf(stderr, "i2cset version %s\n", LM_VERSION);
		exit(0);
	}

	if (argc < flags + 5)
		help();

	i2cbus = strtol(argv[flags+1], &end, 0);
	if (*end || i2cbus < 0 || i2cbus > 0x3f) {
		fprintf(stderr, "Error: I2CBUS argument invalid!\n");
		help();
	}

	address = strtol(argv[flags+2], &end, 0);
	if (*end || address < 0 || address > 0x7f) {
		fprintf(stderr, "Error: Chip address invalid!\n");
		help();
	}

	daddress = strtol(argv[flags+3], &end, 0);
	if (*end || daddress < 0 || daddress > 0xff) {
		fprintf(stderr, "Error: Data address invalid!\n");
		help();
	}

	value = strtol(argv[flags+4], &end, 0);
	if (*end) {
		fprintf(stderr, "Error: Data value invalid!\n");
		help();
	}

	if (argc < flags + 6) {
		fprintf(stderr, "No size specified (using byte-data access)\n");
		size = I2C_SMBUS_BYTE_DATA;
	} else if (argv[flags+5][0] == 'b') {
		size = I2C_SMBUS_BYTE_DATA;
		pec = argv[flags+5][1] == 'p';
	} else if (argv[flags+5][0] == 'w') {
		size = I2C_SMBUS_WORD_DATA;
		pec = argv[flags+5][1] == 'p';
	} else {
		fprintf(stderr, "Error: Invalid mode!\n");
		help();
	}

	if (argc >= flags + 7) {
		vmask = strtol(argv[flags+6], &end, 0);
		if (*end || vmask == 0) {
			fprintf(stderr, "Error: Data value mask invalid!\n");
			help();
		}
	}

	if (value < 0
	 || (size == I2C_SMBUS_BYTE_DATA && value > 0xff)
	 || (size == I2C_SMBUS_WORD_DATA && value > 0xffff)) {
		fprintf(stderr, "Error: Data value out of range!\n");
		help();
	}

	file = open_i2c_dev(i2cbus, filename, 0);
	if (file < 0) {
		exit(1);
	}

	/* check adapter functionality */
	if (ioctl(file, I2C_FUNCS, &funcs) < 0) {
		fprintf(stderr, "Error: Could not get the adapter "
		        "functionality matrix: %s\n", strerror(errno));
		exit(1);
	}

	switch (size) {
	case I2C_SMBUS_BYTE_DATA:
		if (!(funcs & I2C_FUNC_SMBUS_WRITE_BYTE_DATA)) {
			fprintf(stderr, "Error: Adapter for i2c bus %d does "
			        "not have byte write capability\n", i2cbus);
			exit(1);
		}
		break;

	case I2C_SMBUS_WORD_DATA:
		if (!(funcs & I2C_FUNC_SMBUS_WRITE_WORD_DATA)) {
			fprintf(stderr, "Error: Adapter for i2c bus %d does "
			        "not have word write capability\n", i2cbus);
			exit(1);
		}
		break;
	}

	if (set_slave_addr(file, address, force) < 0)
		exit(1);

	if (!yes) {
		int dont = 0;

		fprintf(stderr, "WARNING! This program can confuse your I2C "
		        "bus, cause data loss and worse!\n");

		if (address >= 0x50 && address <= 0x57) {
			fprintf(stderr, "DANGEROUS! Writing to a serial "
			        "EEPROM on a memory DIMM\nmay render your "
			        "memory USELESS and make your system "
			        "UNBOOTABLE!\n");
			dont = 1;
		}

		fprintf(stderr, "I will write to device file %s, chip address "
		        "0x%02x, data address\n0x%02x, data 0x%02x%s, mode "
		        "%s.\n", filename, address, daddress, value,
			vmask ? " (masked)" : "",
			size == I2C_SMBUS_BYTE_DATA ? "byte" : "word");
		if (pec)
			fprintf(stderr, "PEC checking enabled.\n");

		fprintf(stderr, "Continue? [%s] ", dont ? "y/N" : "Y/n");
		fflush(stderr);
		if (!user_ack(!dont)) {
			fprintf(stderr, "Aborting on user request.\n");
			exit(0);
		}
	}

	if (vmask) {
		int oldvalue;

		if (size == I2C_SMBUS_WORD_DATA) {
			oldvalue = i2c_smbus_read_word_data(file, daddress);
		} else {
			oldvalue = i2c_smbus_read_byte_data(file, daddress);
		}

		if (oldvalue < 0) {
			fprintf(stderr, "Error: Failed to read old value\n");
			exit(1);
		}

		value = (value & vmask) | (oldvalue & ~vmask);

		if (!yes) {
			fprintf(stderr, "Old value 0x%0*x, write mask "
				"0x%0*x: Will write 0x%0*x to register "
				"0x%02x\n",
				size == I2C_SMBUS_WORD_DATA ? 4 : 2, oldvalue,
				size == I2C_SMBUS_WORD_DATA ? 4 : 2, vmask,
				size == I2C_SMBUS_WORD_DATA ? 4 : 2, value,
				daddress);

			fprintf(stderr, "Continue? [Y/n] ");
			fflush(stderr);
			if (!user_ack(1)) {
				fprintf(stderr, "Aborting on user request.\n");
				exit(0);
			}
		}
	}

	if (pec) {
		if (ioctl(file, I2C_PEC, 1) < 0) {
			fprintf(stderr, "Error: Could not set PEC: %s\n",
			        strerror(errno));
			exit(1);
		}
		if (!(funcs & (I2C_FUNC_SMBUS_HWPEC_CALC | I2C_FUNC_I2C))) {
			fprintf(stderr, "Warning: Adapter for i2c bus %d does "
			        "not seem to actually support PEC\n", i2cbus);
		}
	}

	e1 = 0;
	if (size == I2C_SMBUS_WORD_DATA) {
		res = i2c_smbus_write_word_data(file, daddress, value);
	} else {
		res = i2c_smbus_write_byte_data(file, daddress, value);
	}
	if (res < 0) {
		fprintf(stderr, "Warning - write failed\n");
		e1++;
	}

	if (pec) {
		if (ioctl(file, I2C_PEC, 0) < 0) {
			fprintf(stderr, "Error: Could not clear PEC: %s\n",
				strerror(errno));
			close(file);
			exit(e1);
		}
	}

	if (size == I2C_SMBUS_WORD_DATA) {
		res = i2c_smbus_read_word_data(file, daddress);
	} else {
		res = i2c_smbus_read_byte_data(file, daddress);
	}
	close(file);

	if (res < 0) {
		fprintf(stderr, "Warning - readback failed\n");
		e1++;
	} else
	if (res != value) {
		e1++;
		fprintf(stderr, "Warning - data mismatch - wrote "
		        "0x%0*x, read back 0x%0*x\n",
		        size == I2C_SMBUS_WORD_DATA ? 4 : 2, value,
		        size == I2C_SMBUS_WORD_DATA ? 4 : 2, res);
	} else {
		fprintf(stderr, "Value 0x%0*x written, readback matched\n",
		        size == I2C_SMBUS_WORD_DATA ? 4 : 2, value);
	}

	exit(e1);
}
